### PR TITLE
ESQL: Reorganize optimizer rules

### LINKED ISSUE #111317: ESQL: remove cycle between LogicalPlanOptimizer and its rules
### Description

As part of a [code refactoring](https://github.com/elastic/elasticsearch/pull/109408), rules where moved out of the LogicalPlanOptimizer (LPO) which contains just a number of static utility methods for naming and some static methods shared used by some rules.
This is problematic since it creates a cycle between the LogicalPlanOptimizer and the rules that refer back to it, such as PushDownEval.

Opening LPO one has a nice javadoc explaining its behavior followed by methods like temporaryName, toString .. before getting to `optimize` which is the core function. 
Or reading about `pushgeneratingPlanPostProjectAndOrderBy` which is relevant to 3 different rules in a different package and has no relationship to LPO.

This should be fixed by:
- [ ] moving the static utility methods around naming introduced [here](https://github.com/elastic/elasticsearch/pull/108360) in a separate class. Could be the existing Expressions, Strings or a separate one inside the rules package.
- [ ] move the static methods used by certain rules inside their package. Define them on one rule and then refer to that from the other rules.
- [ ] while at it, move the logical rules under their own package in "optimizer/rules/logical"

Rules were defined in the same class to convey the invisible dependencies between them - the rules are independent yet they transform the plan that trigger the other rules. This 'global view' is not longer there (due to the class size) with the developer having to jump across classes - having to go back and forth is wasted cognitive effort that we should strive to eliminate.

### PR DESCRIPTION
**Pull request by alex-spies**
Closes https://github.com/elastic/elasticsearch/issues/111317

Organize the optimizer rules consistently for all 4 optimizers (logical, physical, local logical, local physical) and move helper methods meant for optimizer rules out of the optimizers into the relevant rules or into helper classes. Make the optimizers now only contain code relevant to the actual optimizer definition, i.e. which batches of rules they use.

Also, consolidate the 2 nearly identical logical `ParameterizedRule`s into one.

New package structure:
```
package org.elasticsearch.xpack.esql.optimizer
└── rules
    ├── logical
    │   ├── local
    └── physical
        └── local
```
New helper classes:
```
- org.elasticsearch.xpack.esql.optimizer.rules.logical.pushdown.PushDownUtils
- org.elasticsearch.xpack.esql.optimizer.rules.logical.TemporaryNameUtils
- org.elasticsearch.xpack.esql.optimizer.rules.physical.local.LucenePushDownUtils
```

### CODE REVIEW

**Comment by costin** [comment]
Thanks for looking into this.
Quick feedback: at the logical level, the pushdown and local are on the same level which is for me confusing as local calls rules form the parent package as well; I suggest getting rid of the pushdown package and moving the place next to the rest of the other rules as there's nothing special about them.
`lucenepushdown` - quite the mouthful. At least at the moment it's the only time of pushdown possible in a local node and conceptually related to the other rules. I suggest to move these rules under local.

So the final structure looks something like:
```
package org.elasticsearch.xpack.esql.optimizer
└── rules
    ├── logical
    │   ├── local
    └── physical
        └── local
```

+1 on the separating the utils - even better make them protected.

**Comment by alex-spies** [comment]
@costin, the structure is now the one you suggested + I made the new helpers package private.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/DependencyConsistency.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.elasticsearch.xpack.esql.core.expression.AttributeSet;
+import org.elasticsearch.xpack.esql.core.expression.NameId;
+import org.elasticsearch.xpack.esql.plan.QueryPlan;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.esql.common.Failure.fail;
+
+public class DependencyConsistency<P extends QueryPlan<P>> {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/DependencyConsistency.java:21]: The name should indicate the role of this class - checking the consistency of dependencies. Right now it sounds like a DTO class which contains the dependencies consistency.

How about something more mundane but clearer such as `DependencyChecker`, `PlanStateChecker` or `ValidatePlanState`?

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java [hunk 2]
```diff
@@ -153,660 +75,4 @@ protected List<Batch<PhysicalPlan>> rules(boolean optimizeForEsSource) {
     protected List<Batch<PhysicalPlan>> batches() {
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:75]: Maybe a personal preference or looking at this code for long enough to start having a habit: I find it more logical to see the `batches()` definition first and then what a batch contains. In other words, I prefer `batches` be defined before `rules`.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 3]
```diff
@@ -119,30 +93,6 @@ public LogicalPlanOptimizer(LogicalOptimizerContext optimizerContext) {
         super(optimizerContext);
     }
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:95]: For `LogicalPlanOptimizer`, my personal preference is to see these three+ methods in this order: `batches()`, `rules()`, `substitutions()`, `operators()`..... I find it easier to follow the logical order of rules execution.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/DependencyConsistency.java (hunk 1 above)
Comment (astefan) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/DependencyConsistency.java:21]: This class was just extracted from an already existent one, but it wouldn't hurt a rename tbh.

Thinking about the name, I realized we don't have a javadoc for it to explain its intention. How about adding a sentence there?

Leaving the javadoc aside and just looking at the code, I am throwing another name suggestion in the bucket, but **up to @alex-spies to make the change or not altogether**: `PlanConsistencyChecker` (the reasoning behind the name is that this is a checker class and does something, an action; it looks at the "inputs" and "outputs" of each node plan if there is an inconsistency there it spits out an error, thus the `PlanConsistency` part). My 2c.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/OptimizerRules.java [hunk 4]
```diff
@@ -69,4 +70,24 @@ public enum TransformDirection {
         UP,
         DOWN
     }
+
+    public abstract static class ParameterizedOptimizerRule<SubPlan extends LogicalPlan, P> extends ParameterizedRule<
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/OptimizerRules.java:74]: 👍

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/TempNameUtils.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical;
+
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.NameId;
+import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.AggregateFunction;
+
+class TempNameUtils {
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/TempNameUtils.java:16]: Would be too much to name this one `TemporaryNameUtils`? (all its functions use the full `temporary` word and seems more professional to have the class named with a full word instead as well; also, it's not a big name for a class and I don't see why we should shorten its name like this)

**Review by astefan — APPROVED** [review]
LGTM. Thanks, Alex.

I've left some comments mostly spawning from personal preferences and experience. Feel free to adopt or ignore them.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/DependencyConsistency.java (hunk 1 above)
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/DependencyConsistency.java]: I went with Andrei's suggestion :)

**Comment by alex-spies** [comment]
Thanks for the speedy reviews @astefan and @costin !

**Pull request merged**